lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -363,7 +463,59 @@ LookupTypeResult TypeChecker::lookupMemberType(DeclContext *dc,
   if (options.contains(NameLookupFlags::IgnoreAccessibility))
     subOptions |= NL_IgnoreAccessibility;
 
-  if (!dc->lookupQualified(type, name, subOptions, this, decls))
+  auto lookupSuccess = dc->lookupQualified(type, name, subOptions, this, decls);
+
+  // If the requested type is a CodingKeys type on something which may require
+  // Encodable/Decodable synthesis, it may not have been synthesized yet. If
+  // this is the case, force synthesis and attempt lookup again.
+  if (!lookupSuccess && name == Context.Id_CodingKeys) {
```

Comment:
The core code here looks okay, but again I'd like this to be triggered by the name lookup code in the AST library via a `LazyResolver` callback.